### ISSUE TITLE
1p motion correction issues

### ISSUE THREAD

**Issue by donglabimaging** [flatironinstitute/CaImAn#801]
- **Tell us a bit about your setup**: 
 1. Operating system (Linux/macOS/Windows): 
Windows 10 
 
 3. Working environment (Python IDE/Jupyter Notebook/other):
Anaconda
 
 4. Which of the demo scripts you're using for your analysis (if applicable):
 demo_pipeline_cnmfE

 5. CaImAn version*: 
1.8.5 

 6. CaImAn installation process (`pip install .`/`pip install -e .`/conda): 
conda

- **Describe the issue that you are experiencing**
I have been using the demo_pipeline_cnmfE to extract Ca2+ traces from our 1photon images.  The size of our image is 195 x 195.  Upon inspection, the post motion correction video look the same as the original one.  The parameter max_shifts is changed from (3, 3) to (10, 10).  I have tried other values e.g. 25, 30, 48, etc.  The images are still not motion corrected.

motion correction parameters:
`motion_correct = True    # flag for performing motion correction`
`pw_rigid = False         # flag for performing piecewise-rigid motion correction (otherwise just rigid)`
`gSig_filt = (3, 3)       # size of high pass spatial filtering, used in 1p data`
`max_shifts = (10, 10)      # maximum allowed rigid shift`
`strides = (48, 48)       # start a new patch for pw-rigid motion correction every x pixels`
`overlaps = (24, 24)      # overlap between pathes (size of patch strides+overlaps)`
`max_deviation_rigid = 3  # maximum deviation allowed for patch with respect to rigid shifts`
`border_nan = 'copy'      # replicate values along the boundaries`

I used the following code to load the mmap file, inspect the video, and save the post motion correction movie as a tiff file.
`m_rig = cm.load(mc.mmap_file)`
`bord_px_rig = np.ceil(np.max(mc.shifts_rig)).astype(np.int)`
`m_rig.resize(1, 1, 0.2).play(
    q_max=99.5, fr=20, magnification=1, bord_px = 0*bord_px_rig)`
`m_rig.save(os.path.join(os.path.dirname(fnames[0]),'IAMHERE.tif'))`

I tried to look up previous issues, but it seems that no one is having issues with motion correction.  This [LINK](https://drive.google.com/drive/folders/1UGG9BuNKqcI0c_hyZ6Smt9oXuq9BKFq-?usp=sharing) contains the original video (1000 frames in a avi file) and the post motion correction output (saved as tif).  Any suggestions on how to improve the motion correction would be greatly appreciated.  Thank you!

**Comment by EricThomson** [comment]
I haven't looked at your movie, but I had to tweak `gsig_filt` to get it working on our data. `(7,7)` worked well, but it depends on your data.

**Comment by donglabimaging** [comment]
@EricThomson Hi Thanks for the input.  I tried tweaking gsig_filt and it still did not do anything.  I will change it more dramatically and see what happens.

**Comment by donglabimaging** [comment]
@EricThomson I tried a range of `gsig_filt` values (7-30).  It is not still not correcting the motion of the neurons.  I wonder if it has to do with the fact that the cells are in the middle of the view.  For instance, the edge of the lens is stationary and the algorithm is notpicking up the movement.

**Comment by donglabimaging** [comment]
@EricThomson I may have canceled sharing the folder somehow.  Here is the [LINK](https://drive.google.com/drive/folders/10Nlapj_gPGWpA0YKFC4dpYoZa0FfAmip?usp=sharing).  Please let me know if you have trouble viewing the files.

**Comment by donglabimaging** [comment]
I have tried the piecewise correction `pw_rigid = True`.  Although it is not working for our data at the moment, I have incorporated parts of the motion correction demo into our cnmf-e.  I will keep trying until you have time to look at it.  Thank you.

**Comment by donglabimaging** [comment]
It is working now.

**Comment by EricThomson** [comment]
What worked?

**Comment by donglabimaging** [comment]
I played around with the parameters and the motion correction is working
now.  Thanks for the discussion.

Allen

**Comment by donglabimaging** [comment]
@EricThomson Turns out it did not work.  When I said it worked, I have mistakenly imported a video recorded from a different scope.  I am sorry for the confusion.

This link contains the video we would like to analyze [LINK](https://drive.google.com/drive/folders/1UGG9BuNKqcI0c_hyZ6Smt9oXuq9BKFq-?usp=sharing).

I have tried changing the `gsig_filt` from 3-15 and the shift looks best at ~[6, 6] or [7,7].

`gsig_filt = (6, 6)`
![cropped_filt6_shift6_strides48_overlaps24](https://user-images.githubusercontent.com/50585879/92780310-f75a5300-f370-11ea-8e11-c9a7461a3fd2.png)
`gsig_filt = (7, 7)`
![cropped_filt7_shift10_strides48_overlaps24](https://user-images.githubusercontent.com/50585879/92779284-e8bf6c00-f36f-11ea-973e-3492dd80e2be.png)
However, the movements are still there post motion correction.
Code for extracting and inspecting the motion corrected video:
`m_rig = cm.load(mc.mmap_file)`
`bord_px_rig = np.ceil(np.max(mc.shifts_rig)).astype(np.int)`
`m_rig.resize(1, 1, 0.2).play(q_max=99.5, fr=60, magnification=1, bord_px = 0*bord_px_rig)`


I have also tried running **piecewise registration**, but the movements are still there.

Code for performing piecewise registration and inspect the videos.
`mc.pw_rigid = True`  # turn the flag to True for pw-rigid motion correction
`mc.template = mc.mmap_file`  # use the template obtained before to save in computation (optional)
`mc.motion_correct(save_movie=True, template=mc.total_template_rig)`
`m_els = cm.load(mc.fname_tot_els)`
`m_els.resize(0.5, 0.5, downsample_ratio).play(q_max=99.5, fr=60, magnification=2,bord_px = bord_px_rig)`
`cm.concatenate([m_orig.resize(.5, .5, downsample_ratio) - mc.min_mov*mc.nonneg_movie,
                m_rig.resize(.5, .5, downsample_ratio), m_els.resize(
            .5, .5, downsample_ratio)], axis=2).play(fr=60, q_max=99.5, magnification=2, bord_px=bord_px_rig)`

I would really appreciate it if you could provide some comments and suggestions.  Thank you.

**Comment by GarrettBlair** [comment]
I took a look at your video. With the following parameters, I was able to reduce the motion at least. It's not perfect, but its better and should work for source extraction well enough. I'd guess the issue is because the dirt on the lens could be throwing off the motion correction. Here's a link to the motion corrected video: [cropped_MC.tiff](https://drive.google.com/file/d/1d5kUJ11dlUTYLPsuG2o-iOE2s8uTCqKm/view?usp=sharing )

    ```
    mc_dict = {
        'fnames': datadir + 'cropped.tif',
        'fr': 20, # not sure for your data
        'decay_time': .5,
        'pw_rigid': True,
        'max_shifts': (10,10),
        'gSig_filt': (8,8),
        'strides': (32, 32),
        'overlaps': (16, 16),
        'max_deviation_rigid': 3,
        'border_nan': 'copy',
    }
```

**Comment by donglabimaging** [comment]
@GarrettBlair Thank you for your help!  I was not able to access the cropped_MC.tiff link just yet, but I did run the motion correction with the suggested parameter.  I have indeed tried similar parameters and yielded similar results.  I will try to clean the scope and get rid of the dust particles.  Would the motion correction perform even better if the 'strides' and 'overlaps' are further reduced?  Thank you.
